mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusConcurrentTest.java

```diff
@@ -0,0 +1,285 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.events;
+
+import static com.jayway.awaitility.Awaitility.await;
+import static org.apache.james.mailbox.events.EventBusContract.ALL_KEYS;
+import static org.apache.james.mailbox.events.EventBusContract.EVENT;
+import static org.apache.james.mailbox.events.EventBusContract.KEY_1;
+import static org.apache.james.mailbox.events.EventBusContract.KEY_2;
+import static org.apache.james.mailbox.events.EventBusContract.KEY_3;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.james.backend.rabbitmq.RabbitMQConnectionFactory;
+import org.apache.james.backend.rabbitmq.RabbitMQExtension;
+import org.apache.james.event.json.EventSerializer;
+import org.apache.james.mailbox.model.TestId;
+import org.apache.james.mailbox.model.TestMessageId;
+import org.apache.james.util.concurrency.ConcurrentTestRunner;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.testcontainers.shaded.com.google.common.collect.ImmutableList;
+import org.testcontainers.shaded.com.google.common.collect.ImmutableSet;
+
+import com.jayway.awaitility.core.ConditionFactory;
+
+class RabbitMQEventBusConcurrentTest {
+
+    private static class RabbitMQEventBusDeclarationExtension implements BeforeEachCallback, AfterEachCallback {
+
+        private EventSerializer eventSerializer;
+        private RabbitMQConnectionFactory connectionFactory;
+        private RoutingKeyConverter routingKeyConverter;
+
+        @Override
+        public void beforeEach(ExtensionContext context) throws Exception {
+            rabbitMQExtension.beforeEach(context);
+
+            connectionFactory = rabbitMQExtension.getConnectionFactory();
+
+            TestId.Factory mailboxIdFactory = new TestId.Factory();
+            routingKeyConverter = RoutingKeyConverter.forFactories(new MailboxIdRegistrationKey.Factory(mailboxIdFactory));
+            eventSerializer = new EventSerializer(mailboxIdFactory, new TestMessageId.Factory());
+        }
+
+        @Override
+        public void afterEach(ExtensionContext context) throws Exception {
+            rabbitMQExtension.afterEach(context);
+            ALL_LISTENERS.forEach(listener -> listener.clear());
```

Comment:
I prefer re-creating things in the `setUp` rather than having to clear things up - this is what we commonly do with in memory stuff.

I undestand that you do have to deal with an interface. Maybe we can move these listeners as field of this extension, and only alocate them in @BeforeEach ?

Reply:
Moreover `ALL_LISTENERS` is reseted despite this is out of the scope of this extension which clearly is a hack...

Reply:
I extracted that extension into another class file. Because moving all constants from test class to extension class making compilation error. IDK why, but separate them into different file class resolve this.